velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -509,292 +511,413 @@ void castToJsonFromRow(
   });
 }
 
-// Write object to writer at the current offset.
-template <TypeKind kind>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped(
-    const folly::dynamic& /*object*/,
-    exec::GenericWriter&
-    /*writer*/) {
-  VELOX_NYI(
-      "Casting from JSON to {} is not supported.", TypeTraits<kind>::name);
+template <typename T>
+simdjson::simdjson_result<T> fromString(const std::string_view& s) {
+  auto ans = folly::tryTo<T>(s);
+  if (ans.hasError()) {
+    return simdjson::INCORRECT_TYPE;
+  }
+  return std::move(*ans);
 }
 
-// Forward declarations.
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::ARRAY>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer);
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::MAP>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer);
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::ROW>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer);
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::VARCHAR>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer) {
-  if (isJsonType(writer.type())) {
-    // Sort keys to match Presto's behavior.
-    folly::json::serialization_opts opts;
-    opts.sort_keys = true;
-    writer.castTo<Varchar>().append(folly::json::serialize(object, opts));
-  } else if (object.isBool()) {
-    writer.castTo<Varchar>().append(object.asBool() ? "true" : "false");
-  } else {
-    writer.castTo<Varchar>().append(object.asString());
+template <typename To, typename From>
+simdjson::error_code convertIfInRange(From x, exec::GenericWriter& writer) {
+  static_assert(std::is_signed_v<From> && std::is_signed_v<To>);
+  static_assert(std::is_integral_v<To> || !std::is_integral_v<From>);
+  if constexpr (!std::is_same_v<To, From>) {
+    constexpr From kMin = std::numeric_limits<To>::lowest();
+    constexpr From kMax = std::numeric_limits<To>::max();
+    if (!(kMin <= x && x <= kMax)) {
+      return simdjson::NUMBER_OUT_OF_RANGE;
+    }
   }
+  writer.castTo<To>() = x;
+  return simdjson::SUCCESS;
 }
 
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::BOOLEAN>(
-    const folly::dynamic& object,
+template <TypeKind kKind>
```

Comment:
kKind -> kind

Reply:
This is the naming convention to show it is a compile time constant: https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#naming-conventions

Reply:
Hmm... this doesn't apply to type parameters. At least, this is not what we do currently in the codebase.

Reply:
But type parameters are static constants

Reply:
That's true, hence, if we were following kXxx naming then all type parameters would have to be named like that: kTo, kFrom. We don't do that because this naming convention doesn't apply to types.

Reply:
`From` and `To` are `typename` parameter, so just capitalize them is enough.  The `k` prefix rule only applies to type parameter that is not `typename` or `class`